src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,482 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
     }
 };
 
+class fetch_worker {
+public:
+    // Passed from the coordinator shard to fetch workers.
+    // Contains either references to objects local to the fetch worker shard or
+    // copies of data from the coordinator shard.
+    struct shard_local_fetch_context {
+        // True if the results of this fetch will be read on a foreign shard.
+        bool foreign_read;
+        size_t bytes_left;
+        // Specifies the minimum number of bytes this sub-fetch should read
+        // before returning.
+        size_t min_bytes;
+        // If set then the sub-fetch should return by the specified time_point .
+        std::optional<model::timeout_clock::time_point> deadline;
+        // The fetch sub-requests of partitions local to the shard this worker
+        // is running on.
+        std::vector<ntp_fetch_config> requests;
+
+        // References to services local to the shard this worker is running on.
+        // They are protected from deletion by the coordinator.
```

Comment:
This is hard to understand. When you talk about "deletion" are you talking about the service itself, or the reference? The reference is inside this struct so its lifetime should be the same as the whole struct.

If the service itself, how does the coordinator prevent deletion of services on the worker?

Reply:
I need to double check, but with sharded services the `stop()` function on every shard needs to complete before the service is deleted on any shard. The coordinator not returning indirectly prevents a gate from closing and hence the `stop()` function on those services from completing. And since the coordinator won't return until the workers complete it prevents the deletion of those services on the worker shards as well.

Reply:
Thanks for the explanation. So bascailly the lifetime of all workers is scoped within the lifetime of the request, and the request lifetimes are presumamably scoped within the connection, and in turn connections are scoped within the service, or something like that?

Reply:
Yeah, it's a long easily broken chain, but it seems to work if we're careful.